extensions/socket/internal.m

```diff
@@ -0,0 +1,657 @@
+
+    if (lua_type(L, 2) == LUA_TFUNCTION) {
+        lua_pushvalue(L, 2);
+        asyncSocket.readCallback = [skin luaRef:refTable];
+    }
+
+    lua_pushvalue(L, 1);
+    return 1;
+}
+
+/// hs.socket:setTimeout(timeout) -> self
+/// Method
+/// Sets the timeout for the socket operations. If the timeout value is negative, the operations will not use a timeout
+///
+/// Parameters:
+///  * timeout - A number containing the timeout duration, in seconds
+///
+/// Returns:
+///  * The [`hs.socket`](#new) object
+///
+static int socket_setTimeout(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TNUMBER, LS_TBREAK];
+
+    HSAsyncSocket* asyncSocket = getUserData(L, 1);
+    NSTimeInterval timeout = lua_tonumber(L, 2);
+    asyncSocket.timeout = timeout;
+
+    lua_pushvalue(L, 1);
+    return 1;
+}
+
+/// hs.socket:startTLS([verify][, peerName]) -> self
+/// Method
+/// Secures the socket with TLS. The socket will disconnect immediately if TLS negotiation fails
+///
+/// Parameters:
+///  * verify - An optional boolean that, if `false`, allows TLS handshaking with servers with self-signed certificates and does not evaluate the chain of trust. Defaults to `true`
+///  * peerName - An optional string containing the fully qualified domain name of the peer to validate against — for example, `store.apple.com`. It should match the name in the X.509 certificate given by the remote party. See notes below
+///
+/// Returns:
+///  * The [`hs.socket`](#new) object
+///
+/// Notes:
+/// * IMPORTANT SECURITY NOTE:
+/// The default settings will check to make sure the remote party's certificate is signed by a
+/// trusted 3rd party certificate agency (e.g. verisign) and that the certificate is not expired.
+/// However it will not verify the name on the certificate unless you
+/// give it a name to verify against via the kCFStreamSSLPeerName key.
+/// The security implications of this are important to understand.
+/// Imagine you are attempting to create a secure connection to MySecureServer.com,
+/// but your socket gets directed to MaliciousServer.com because of a hacked DNS server.
+/// If you simply use the default settings, and MaliciousServer.com has a valid certificate,
+/// the default settings will not detect any problems since the certificate is valid.
+/// To properly secure your connection in this particular scenario you
+/// should set the kCFStreamSSLPeerName property to "MySecureServer.com".
+///
+static int socket_startTLS(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBOOLEAN|LS_TOPTIONAL, LS_TSTRING|LS_TOPTIONAL, LS_TBREAK];
+
+    HSAsyncSocket* asyncSocket = getUserData(L, 1);
+    NSDictionary *tlsSettings = nil;
+
+    if (lua_type(L, 2) == LUA_TBOOLEAN && lua_toboolean(L, 2) == false) {
+        tlsSettings = @{@"GCDAsyncSocketManuallyEvaluateTrust": @YES};
+    }
+
+    if (lua_type(L, 3) == LUA_TSTRING) {
```

Comment:
I'm not sure this is right - the docstrings above suggest that both arguments are optional, so `L,2` might also be `LUA_TSTRING` if the user is omitting the `verify` argument. This would also need the `checkArgs` signature to change to `LS_TBOOLEAN|LS_TSTRING|LS_TOPTIONAL`.